design-documents/service-isolation/inter-service-communication.md

```diff
@@ -0,0 +1,398 @@
+
+When calling the invoker only interface names and methods names will be given (because those are the service declarations)
+and it would be hard to determine a manually created URLs for endpoints assigned to requested services. Also, except for
+services written using other technologies than Magento, all the requests to services will generated automatically by
+invoker. Therefore there is no need to assign URLs to service contracts manually - they can be generated automatically
+based on interfaces/methods names.
+
+HTTP methods in RESTful API are used to provide self-documentation for developers writing clients and for browsers to
+adjust their behaviour when processing requests to RESTful APIs. They also can be used for caching rules for web
+servers/proxies like Varnish. In services communications requests are being generated automatically, they don't
+come through browsers and caching rules for servers must be assessed on a case-to-case basis and cache needs to be
+carefully invalidated to avoid breaking services' business logic, therefore there is no need in HTTP methods for
+services communications and invoker will always send POST requests to remote services. Services will accept any
+HTTP method though so developers of services not based on Magento can send requests with methods they feel
+suite a requests best.
+
+While URLs for services will be autogenerated and POST always used to make it easier for Magento's invoker it will still
+be possible for developers to write services without using Magento (or even PHP) - they would be able to get see URLs
+for certain contracts with a CLI utility and while using POST for every request looks odd, functionally it wouldn't matter.
+So developers could contact Magento services just like any other microservice with RESTful web API.
+  
+The way exceptions thrown by services are serialized right now for RESTful web API needs to be changed for inter-service endpoints.
+Exceptions are serialized without their type information provided - this info is crutial for services to communicate with each other so
+the type information will be added to the output.
+ 
+To make it easier for debuging and to perhaps later to be used for data consistency _X-Magento-Correlation-Id_ header will
+be added to requests to remote services with ID generated on BFF before a 1st request to a service. The ID will be
+forwarded between services to identify a single operation.  
+ 
+##### Retries mechanism
+The invoker implementation based on RESTful web APIs will offer default retries mechanism, but developers will be able to disable it
+by using _di.xml_ (see Invoker API and SPI [here](invoker.md)) for when they would want to delegate this responsibility
+to a service mash or a balancer. 
+The timeout and tries limit will be configured via di.xml (see the invoker API and SPI [here](invoker.md)).
+This default implementation may be used by developers who choose to avoid service mashes or for development environments.
```

Comment:
I think a retries mechanism should a part of the infrastructure, not an invoker's part. For development purposes, probably, you don't need retries at all. And not sure if we are building a system where services could be deployed and located separately, how we can avoid the usage of some kind of service mesh.
For development purposes, we should provide an environment not only for 3rd party developers but for core teams also.